File: php/phpsearch/filetypes.php
<?php

require_once __DIR__ . '/autoload.php';

class FileTypes {
    function __construct() {
        $this->file_type_map = $this->get_file_type_map();
    }

    private function get_file_type_map() {
        $file_type_map = array();
        $filetypespath = FileUtil::expand_user_home_path(Config::FILETYPESPATH);
        if (file_exists($filetypespath)) {
            $filetypes = simplexml_load_file($filetypespath);
            foreach ($filetypes->filetype as $filetype) {
                $name = sprintf($filetype['name']);
                $exts = preg_split("/\s+/", $filetype->extensions);
                $file_type_map[$name] = $exts;
            }
            $file_type_map['text'] = array_merge($file_type_map['text'],
                $file_type_map['code'], $file_type_map['xml']);
            $file_type_map['searchable'] = array_merge($file_type_map['text'],
                $file_type_map['archive'], $file_type_map['binary']);
        } else {
            throw new Exception('File not found: ' . $filetypespath);
        }
        return $file_type_map;
    }

    public static function from_name(string $name) {
        $uname = strtoupper($name);
        if ($uname == 'TEXT') {
            return FileType::Text;
        }
        if ($uname == 'BINARY') {
            return FileType::Binary;
        }
        if ($uname == 'ARCHIVE') {
            return FileType::Archive;
        }
        if ($uname == 'CODE') {
            return FileType::Code;
        }
        if ($uname == 'XML') {
            return FileType::Xml;
        }
        return FileType::Unknown;
    }

    public function get_filetype(string $file) {
        if ($this->is_text($file)) {
            return FileType::Text;
        }
        if ($this->is_binary($file)) {
            return FileType::Binary;
        }
        if ($this->is_archive($file)) {
            return FileType::Archive;
        }
        if ($this->is_code($file)) {
            return FileType::Code;
        }
        if ($this->is_xml($file)) {
            return FileType::Xml;
        }
        return FileType::Unknown;
    }

    public function is_archive(string $f): bool {
        return in_array(FileUtil::get_extension($f), $this->file_type_map['archive']);
    }

    public function is_binary(string $f): bool {
        return in_array(FileUtil::get_extension($f), $this->file_type_map['binary']);
    }

    public function is_code(string $f): bool {
        return in_array(FileUtil::get_extension($f), $this->file_type_map['code']);
    }

    public function is_text(string $f): bool {
        return in_array(FileUtil::get_extension($f), $this->file_type_map['text']);
    }

    public function is_xml(string $f): bool {
        return in_array(FileUtil::get_extension($f), $this->file_type_map['xml']);
    }

    public function is_searchable(string $f): bool {
        return in_array(FileUtil::get_extension($f), $this->file_type_map['searchable']);
    }

    public function is_unknown(string $f): bool {
        return !$this->is_searchable($f);
    }
}

?>

File: ocaml/mlsearch/src/filetypes.ml
open Core.Std
open Xml

type file_type =
  | Unknown
  | Archive
  | Binary
  | Code
  | Text
  | Xml;;

module StringMap = Map.Make (String)

type t = (string list) StringMap.t

let empty = StringMap.empty

let to_list t = Map.to_alist t

let ws_regex = Re2.Regex.create_exn "\\s+"

let get_extensions file_type_node =
  match (List.hd (Xml.children file_type_node)) with
  | Some ext_node -> (
    match (List.hd (Xml.children ext_node)) with
    | Some exts -> Re2.Regex.split ws_regex (String.strip (Xml.pcdata exts))
    | None -> [])
  | None -> [];;

let get_file_types : (string list) StringMap.t =
  let x = Xml.parse_file (Config.xsearchpath ^ "/shared/filetypes.xml") in
  let file_type_nodes = Xml.children x in
  let fts = List.map file_type_nodes
    ~f:(fun ft -> ((Xml.attrib ft "name"), (get_extensions ft))) in
  StringMap.of_alist_exn fts;;

let is_archive t f : bool =
  List.mem (StringMap.find_exn t "archive") (Fileutil.get_extension f);;

let is_binary t f : bool =
  List.mem (StringMap.find_exn t "binary") (Fileutil.get_extension f);;

let is_code t f : bool =
  List.mem (StringMap.find_exn t "code") (Fileutil.get_extension f);;

let is_text t f : bool =
  let text_types = ["code"; "text"; "xml";] in
  List.exists text_types ~f:(fun tt -> List.mem (StringMap.find_exn t tt) (Fileutil.get_extension f));;

let is_xml t f : bool =
  List.mem (StringMap.find_exn t "xml") (Fileutil.get_extension f);;

let get_file_type t f : file_type =
  if is_text t f then Text
  else if is_binary t f then Binary
  else if is_archive t f then Archive
  else if is_code t f then Code
  else if is_xml t f then Xml
  else Unknown;;
